toolchain/check/impl_lookup.cpp

```diff
@@ -183,79 +181,134 @@ static auto GetInterfaceIdFromConstantId(Context& context, SemIR::LocId loc_id,
+  // of the impl.
+  //
+  // As a shortcut, if the impl's constraint is not symbolic (does not depend on
+  // any generic parameters), then we can determine if we match if the specific
+  // ids match exactly.
+  auto impl_interface_const_id =
+      context.constant_values().Get(impl.constraint_id);
+  if (!impl_interface_const_id.is_symbolic()) {
+    if (impl.interface.specific_id != interface.specific_id) {
+      return SemIR::InstId::None;
+    }
+  }
+
   // This check comes first to avoid deduction with an invalid impl. We use an
   // error value to indicate an error during creation of the impl, such as a
   // recursive impl which will cause deduction to recurse infinitely.
   if (impl.witness_id == SemIR::ErrorInst::SingletonInstId) {
     return SemIR::InstId::None;
   }
+  if (!impl.witness_id.has_value()) {
+    // TODO: Diagnose if the impl isn't defined yet?
+    return SemIR::InstId::None;
+  }
+
+  // The impl may have generic arguments, in which case we need to deduce them
+  // to find what they are given the specific interface query. We use that
+  // specific to map values in the impl to the deduced values.
+  auto specific_id = SemIR::SpecificId::None;
   if (impl.generic_id.has_value()) {
     specific_id = DeduceImplArguments(context, loc_id, impl, type_const_id,
-                                      interface_const_id);
+                                      interface.specific_id);
     if (!specific_id.has_value()) {
       return SemIR::InstId::None;
     }
   }
-  if (!context.constant_values().AreEqualAcrossDeclarations(
-          SemIR::GetConstantValueInSpecific(context.sem_ir(), specific_id,
-                                            impl.self_id),
-          type_const_id)) {
+
+  // The self type of the impl must match the type in the query, or this is an
+  // `impl T as ...` for some other type `T` and should not be considered.
+  auto deduced_self_const_id = SemIR::GetConstantValueInSpecific(
+      context.sem_ir(), specific_id, impl.self_id);
+  if (type_const_id != deduced_self_const_id) {
     return SemIR::InstId::None;
   }
-  if (!context.constant_values().AreEqualAcrossDeclarations(
-          SemIR::GetConstantValueInSpecific(context.sem_ir(), specific_id,
-                                            impl.constraint_id),
-          interface_const_id)) {
-    // TODO: An impl of a constraint type should be treated as implementing
-    // the constraint's interfaces.
+
+  // The impl's constraint is a facet type which it is implementing for the self
+  // type: the `I` in `impl ... as I`. The deduction step may be unable to be
+  // fully applied to the types in the constraint and result in an error here,
+  // in which case it does not match the query.
+  auto deduced_constraint_id =
+      context.constant_values().GetInstId(SemIR::GetConstantValueInSpecific(
+          context.sem_ir(), specific_id, impl.constraint_id));
+  if (deduced_constraint_id == SemIR::ErrorInst::SingletonInstId) {
```

Comment:
Can you give me an example where this triggers?

Reply:
I think I don't understand the comment and an example would help me figure out what I was missing.

Reply:
Yes, there's a test that causes this.

```
interface ComparableWith(T:! type) {}

// This creates a dependency cycle with itself.
impl forall [U:! type, T:! ComparableWith(U)] U as ComparableWith(T) {}

class C {}
class D {}
impl C as ComparableWith(D) {}

fn Compare[T:! type, U:! ComparableWith(T)](t: T, u: U) {}

fn F() {
  Compare({} as C, {} as C);
}
```

The deduction of `Compare`'s `U` finds a cycle in impl lookup, which returns an errorinst for that deduced type. When we apply the deduced specifics to the constraint of the `impl forall`'s constraint `ComparableWith(T)` we get back an error here.